internal/validators/rbac.go

```diff
@@ -0,0 +1,145 @@
+	latestCondition.ValidationType = constants.ValidationTypeRBAC
+	validationResult := &vapitypes.ValidationResult{Condition: &latestCondition, State: &state}
+
+	failures := make([]string, 0)
+
+	for i, set := range rule.Permissions {
+		s.log.V(0).Info("Processing permission set of rule.", "set #", i+1)
+		if err := s.processPermissionSet(set, rule.SecurityPrincipalID, &failures); err != nil {
+			// Code this is returning to will take care of changing the validation result to a
+			// failed validation, using the error returned.
+			return validationResult, err
+		}
+	}
+
+	if len(failures) > 0 {
+		state = vapi.ValidationFailed
+		latestCondition.Failures = failures
+		latestCondition.Message = "Security principal missing one or more required roles."
+		latestCondition.Status = corev1.ConditionFalse
+	}
+
+	return validationResult, nil
+}
+
+// processPermissionSet processes a permission set from the rule.
+//   - set: The set to process.
+//   - principalID: The ID of the security principal to use in the filter. This comes from the rule
+//     the set is part of.
+//   - failures: The list of failures being built up while processing the entire rule. Must be
+//     non-nil.
+func (s *RBACRuleService) processPermissionSet(set v1alpha1.PermissionSet, principalID string, failures *[]string) error {
+
+	foundRoleNames := make(map[string]bool)
+
+	// Get all role assignments that apply to the specified scope where the member of the role
+	// assignment is the specified security principal. In this query, "principalId" must be a UUID,
+	// so this shouldn't have any injection vulnerabilities.
+	//
+	// Note that this also returns role assignments that assign the role because the scope is a
+	// surrounding scope (e.g. the subscription the scope is contained within), not just the scope
+	// itself.
+	filter := ptr.Ptr(url.QueryEscape(fmt.Sprintf("principalId eq '%s'", principalID)))
+	roleAssignments, err := s.api.ListRoleAssignmentsForScope(set.Scope, filter)
+	if err != nil {
+		return fmt.Errorf("failed to get role assignments: %w", err)
+	}
+
+	for _, ra := range roleAssignments {
+		if ra.Properties != nil && ra.Properties.RoleDefinitionID != nil {
+			foundRoleNames[azure_utils.RoleNameFromRoleDefinitionID(*ra.Properties.RoleDefinitionID)] = true
+		}
+	}
+
+	// First, find out whether we need to look the role up by its role name if the user provided
+	// its role name instead of its name.
+	var roleName string
+	role := set.Role
+	if role.Name != nil {
+		roleName = *role.Name
+	} else if role.RoleName != nil {
+		// To do the role name lookup, we need to get all of the role definitions that exist in the
+		// subscription that we're working with. We figure out which subscription we're working with
+		// by using the subscription from the scope of the permission set we're working on.
+		subForLookup, err := azure_utils.RoleAssignmentScopeSubscription(set.Scope)
```

Comment:
I'm pretty sure this is safe to do because as far as I could tell, the role assignment scope will always have a subscription in it. So any values our users will ever configure will have enough data for us to do things this way.

See https://learn.microsoft.com/en-us/azure/role-based-access-control/scope-overview for details.

And I just realized I'm wrong.

![image](https://github.com/spectrocloud-labs/validator-plugin-azure/assets/7719209/86b0aa8a-f801-4c40-be97-c5529af1ea18)

"Management group" is a higher level of scope we might want to add support for.

Reply:
Let's ignore Management group for now.